### PR TITLE
Fc features

### PR DESCRIPTION
**Pull request by cmfcmf**

### CODE REVIEW

File: src/lib/Zikula/Bundle/CoreInstallerBundle/Stage/Upgrade/AjaxUpgraderStage.php [hunk 1]
```diff
@@ -73,20 +73,27 @@ public function getTemplateParams()
                 AjaxInstallerStage::FAIL => __('There was an error regenerating the themes')
             ),
             6 => array(
```
Comment (cmfcmf) [inline src/lib/Zikula/Bundle/CoreInstallerBundle/Stage/Upgrade/AjaxUpgraderStage.php]: Couldn't those keys be removed entirely?

File: src/lib/Zikula/Common/Translator/TranslatorTrait.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * @var \Zikula\Common\Translator\Translator
+     */
+    private $translator;
+    /**
+     * singular translation for modules.
+     *
+     * @param string $msg Message.
+     * @param null $domain
+     * @param null $locale
```
Comment (cmfcmf) [inline src/lib/Zikula/Common/Translator/TranslatorTrait.php]: I don't think this should say `null` but rather string|null

File: src/lib/Zikula/Core/Doctrine/Helper/SchemaHelper.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $metaClasses = array();
+        foreach ($classes as $class) {
+            $metaClasses[] = $this->entityManager->getClassMetadata($class);
+        }
+        try {
+            $this->tool->createSchema($metaClasses);
+        } catch (\PDOException $e) {
+            throw $e;
+        }
```
Comment (cmfcmf) [inline src/lib/Zikula/Core/Doctrine/Helper/SchemaHelper.php]: I don't understand the reason for this catch block

File: src/lib/Zikula/Core/Theme/Engine.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+     * Find the realm in the theme.yml that matches the given path, route or module
+     * Uses regex to match a pattern to one of three possible values
+     *
+     * @todo is there a faster way to do this?
+     * @return int|string
+     */
+    private function setMatchingRealm()
+    {
+        foreach ($this->themeBundle->getConfig() as $realm => $config) {
+            if (!empty($config['pattern'])) {
```
Comment (cmfcmf) [inline src/lib/Zikula/Core/Theme/Engine.php]: I don't like how the pattern is matched against everything - I'd like to see three distinct settings for these.

File: src/lib/legacy/Zikula/Core.php [hunk 5]
```diff
@@ -55,7 +55,7 @@ class Zikula_Core
     /**
      * The core Zikula version number.
      */
-    const VERSION_NUM = '1.4.0';
+    const VERSION_NUM = '1.4.1';
```
Comment (cmfcmf) [inline src/lib/legacy/Zikula/Core.php]: Is this intended now?

File: src/lib/legacy/viewplugins/outputfilter.asseturls.php [hunk 6]
```diff
@@ -28,7 +28,7 @@
  */
 function smarty_outputfilter_asseturls($source, $view)
 {
-    $source = preg_replace_callback('#(href=|src=){1}("|\'){1}([a-zA-Z0-9/\.\-_]+)("|\'){1}#', '_smarty_outputfilter_asseturls', $source);
+    $source = preg_replace_callback('#(href=|src=){1}("|\'){1}([a-zA-Z0-9\/\.\-_]+)("|\'){1}#', '_smarty_outputfilter_asseturls', $source);
```
Comment (cmfcmf) [inline src/lib/legacy/viewplugins/outputfilter.asseturls.php]: How'd you spot this one? :smile: :+1:

File: src/lib/util/PluginUtil.php [hunk 7]
```diff
@@ -160,7 +160,8 @@ public static function loadPlugin($className)
             $plugin->setBooted();
         }
 
-        return $sm->set($serviceId, $plugin);
+        $sm->set($serviceId, $plugin);
+        return $plugin;
```
Comment (cmfcmf) [inline src/lib/util/PluginUtil.php]: Blank line before return :boat:

File: src/system/Zikula/Module/BlocksModule/Api/MenutreeApi.php [hunk 8]
```diff
@@ -143,20 +143,20 @@ public function adminlinks($args)
...
 
-                $modinfo = ModUtil::getInfo(ModUtil::getIdFromName($adminmodule['name']));
-
-                if ($modinfo['type'] == 2 || $modinfo['type'] == 3) {
-                    $menutexturl = ModUtil::url($modinfo['name'], 'admin', 'index');
+                if ($adminmodule['type'] == 2 || $adminmodule['type'] == 3) {
+                    $menutexturl = isset($adminmodule['capabilities']['admin']['url'])
+                        ? $adminmodule['capabilities']['admin']['url']
+                        : $this->get('router')->generate($adminmodule['capabilities']['admin']['route']);
                 } else {
```
Comment (cmfcmf) [inline src/system/Zikula/Module/BlocksModule/Api/MenutreeApi.php]: The Diff looks odd here. Are you sure you didn't use tabs or something?

**Comment by cmfcmf** [comment]
Looking good overall. Thank you @craigh for your work!

File: src/lib/Zikula/Bundle/CoreInstallerBundle/Stage/Upgrade/AjaxUpgraderStage.php (hunk 1 above)
Comment (craigh) [inline src/lib/Zikula/Bundle/CoreInstallerBundle/Stage/Upgrade/AjaxUpgraderStage.php]: probably, but this ensures order.

File: src/lib/Zikula/Core/Theme/Engine.php (hunk 4 above)
Comment (craigh) [inline src/lib/Zikula/Core/Theme/Engine.php]: let's merge as is and then open a discussion on the topic.

Comment (craigh) [inline src/lib/Zikula/Core/Theme/Engine.php]: and I am not clear on what you mean by "three distinct settings"

File: src/lib/legacy/Zikula/Core.php (hunk 5 above)
Comment (craigh) [inline src/lib/legacy/Zikula/Core.php]: yup :wink:

File: src/lib/legacy/viewplugins/outputfilter.asseturls.php (hunk 6 above)
Comment (craigh) [inline src/lib/legacy/viewplugins/outputfilter.asseturls.php]: testing :+1:

File: src/system/Zikula/Module/BlocksModule/Api/MenutreeApi.php (hunk 8 above)
Comment (craigh) [inline src/system/Zikula/Module/BlocksModule/Api/MenutreeApi.php]: yeah. I checked and it is correct. just must be vexing the diff engine.

**Comment by craigh** [comment]
@cmfcmf - some big changes since your review today. :boat:

**Comment by craigh** [comment]
hmm. travis failed to install with this error:

> [Symfony\Component\DependencyInjection\Exception\RuntimeException] A string value must be composed of strings and/or numbers, but found parameter "router.request_context.base_url" of type NULL inside string value "%router.request_context.base_url%/web/jquery/jquery.min.js".

which means the installer is broken for a minute. I guess I will follow up with the *@todo* in that section to fix the error...

**Comment by craigh** [comment]
@cmfcmf @Guite I would really like to merge this now. please voice any objections ASAP.

@cmfcmf - the StyleCI analyzer works better now (see output) but needs adjustment to allow spaces before and after concatenation dots (e.g. `$foo . $bar` instead of `$foo.$bar`) and there are maybe some other adjustments that can be made (again see output).

**Comment by Guite** [comment]
No objections from me. What I've seen looks very promising!

**Comment by Guite** [comment]
about StyleCI @cmfcmf : maybe switch from `recommended` to `symfony` ?
https://styleci.readme.io/docs/configuration
(should probably be discussed in a separate issue)

**Comment by cmfcmf** [comment]
Merge if you like.

About styleci, I just briefly tried it but haven't configured anything yet,
feel free to do so or open an issue.
Axel Guckelsberger <notifications@github.com> schrieb am So., 16. Aug. 2015
um 08:36:

**Pull request merged**